### PR TITLE
VoIP v1 implementation

### PR DESCRIPTION
**Pull request by ganfra**
This PR begins by introducing version 1 of VoIP protocol (Details: matrix-org/matrix-doc#2746)

- Change the version field in all VoIP events to 1
- Add party_id to all VoIP events
- Introduce m.call.select_answer
- Introduce m.call.reject
- Introduce more reason codes to m.call.hangup
- Introduce m.call.negotiate (hold/resume, ice restart...)

It now contains every new features and fixes since the beginning, including : 
- Transfer call
- Use dial-pad in-all for sending DTMF
- Fix audio devices output
- Handle having 2 calls
- Create new tiles in timeline for call

photo_2020-12-02 15 36 45
photo_2020-12-02 15 36 44

### CODE REVIEW

**Review by dbkr — COMMENTED** [review]
This is looking awesome :)

**Comment by bmarty** [comment]
Can you rebase (or merge develop to) the branch, fix issue reported by CI and update the file CHANGES.md please?

**Comment by ganfra** [comment]
> Can you rebase (or merge develop to) the branch, fix issue reported by CI and update the file CHANGES.md please?

Of course I will, but this is just a draft PR :)

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/model/call/CallRejectContent.kt [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        @Json(name = "call_id") override val callId: String,
+        /**
+         * Required. ID to let user identify remote echo of their own events
+         */
+        @Json(name = "party_id") override val partyId: String? = null,
+        /**
+         * Required. The version of the VoIP specification this message adheres to.
+         */
+        @Json(name = "version") override val version: String?
+):CallSignallingContent
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/model/call/CallRejectContent.kt:40]: Bad formatting :)

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/call/CallListenersDispatcher.kt [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import org.matrix.android.sdk.api.session.room.model.call.CallHangupContent
+import org.matrix.android.sdk.api.session.room.model.call.CallInviteContent
+import org.matrix.android.sdk.api.session.room.model.call.CallNegotiateContent
+import org.matrix.android.sdk.api.session.room.model.call.CallRejectContent
+import org.matrix.android.sdk.api.session.room.model.call.CallSelectAnswerContent
+
+/**
+ * Dispatch each method safely to all listeners.
+ */
+class CallListenersDispatcher(private val listeners: Set<CallListener>) : CallListener {
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/call/CallListenersDispatcher.kt:33]: internal

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/call/CallSignalingHandler.kt [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    private fun MxCall.partyIdsMatches(contentSignallingContent: CallSignallingContent): Boolean {
+        return opponentPartyId?.getOrNull() == contentSignallingContent.partyId
+    }
+
+    private fun CallSignallingContent.getCall(): MxCall? {
+        val currentCall = callId?.let {
+            activeCallHandler.getCallWithId(it)
+        }
+        if (currentCall == null) {
+            Timber.v("Call for content: $this is null")
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/call/CallSignalingHandler.kt:202]: I would log only the id and not $this here (privacy?)

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/call/MxCallFactory.kt [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+internal class MxCallFactory @Inject constructor(
+        @DeviceId private val deviceId: String?,
+        private val localEchoEventFactory: LocalEchoEventFactory,
+        private val eventSenderProcessor: EventSenderProcessor,
+        @UserId private val userId: String
+) {
+
+    fun createIncomingCall(roomId: String, senderId: String, content: CallInviteContent): MxCall? {
+        if (content.callId == null) return null
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/call/MxCallFactory.kt:39]: Now I use this simpler form:
`content.callId ?: return null`

File: vector/src/main/res/layout/activity_call.xml [hunk 5]
```diff
@@ -26,70 +36,79 @@
...
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        tools:src="@tools:sample/avatars" />
+
+    <ImageView
+        android:layout_width="20dp"
+        android:layout_height="20dp"
+        android:id="@+id/smallIsHeldIcon"
```
Comment (bmarty) [inline vector/src/main/res/layout/activity_call.xml:54]: please format the file (bad attributes ordering)

File: vector/src/main/res/layout/activity_call.xml [hunk 6]
```diff
@@ -26,70 +36,79 @@
...
         android:layout_width="0dp"
         android:layout_height="wrap_content"
         android:layout_marginStart="@dimen/layout_horizontal_margin"
+        android:layout_marginTop="16dp"
         android:layout_marginEnd="@dimen/layout_horizontal_margin"
         android:gravity="center"
-        android:textColor="?riotx_text_primary"
-        android:textSize="28sp"
+        android:textColor="@color/white"
+        android:textSize="20sp"
```
Comment (bmarty) [inline vector/src/main/res/layout/activity_call.xml:71]: We had a complain that the texts are too small on this screen :/

File: vector/src/main/res/layout/bottom_sheet_call_controls.xml [hunk 7]
```diff
@@ -30,9 +30,18 @@
...
         tools:actionDescription="Front" />
 
+    <im.vector.app.core.ui.views.BottomSheetActionButton
+        android:id="@+id/callControlsToggleHoldResume"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        app:actionTitle="Hold/resume"
+        app:leftIcon="@drawable/ic_call_hold_action"
+        app:tint="?attr/riotx_text_primary"
+        tools:actionDescription="" />
```
Comment (bmarty) [inline vector/src/main/res/layout/bottom_sheet_call_controls.xml:45]: Holding a call is not working yet? (I tried with current Element Web)

File: vector/src/main/res/layout/item_verification_action.xml [hunk 8]
```diff
@@ -60,7 +60,7 @@
         app:layout_constraintStart_toStartOf="@+id/itemVerificationActionTitle"
         app:layout_constraintTop_toBottomOf="@+id/itemVerificationActionTitle"
         tools:text="For maximum security, do this in person"
-        tools:visibility="visible" />
+        />
```
Comment (bmarty) [inline vector/src/main/res/layout/item_verification_action.xml:63]: Why removing the tools:visibility? Please restore

**Review by bmarty — CHANGES_REQUESTED** [review]
Amazing work!
Some remarks, and:
- Holding call does not work yet?
- I had a crash (not reproduced so far) when trying to start a video call. Log: https://rageshakes.proxy.matrix.org/api/listing/2020-12-10/133036/logs-0000.log.gz search for `ConcurrentModificationException`
Please also update the CHANGES.md file :)

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/model/call/CallAnswerContent.kt [hunk 9]
```diff
@@ -27,16 +27,20 @@ data class CallAnswerContent(
         /**
          * Required. The ID of the call this event relates to.
          */
-        @Json(name = "call_id") val callId: String,
+        @Json(name = "call_id") override val callId: String,
+        /**
+         * Required. ID to let user identify remote echo of their own events
+         */
+        @Json(name = "party_id") override val partyId: String? = null,
```
[the remark this answers is not in the document]
Comment (ganfra) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/model/call/CallAnswerContent.kt:34]: Yup

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/model/call/CallAnswerContent.kt [hunk 10]
```diff
@@ -27,16 +27,20 @@ data class CallAnswerContent(
...
          * Required. The session description object
          */
         @Json(name = "answer") val answer: Answer,
         /**
-         * Required. The version of the VoIP specification this messages adheres to. This specification is version 0.
+         * Required. The version of the VoIP specification this messages adheres to.
          */
-        @Json(name = "version") val version: Int = 0
-) {
+        @Json(name = "version") override val version: String?
```
[the remark this answers is not in the document]
Comment (ganfra) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/model/call/CallAnswerContent.kt:42]: I prefer so we don't have any bad surprise

File: vector/build.gradle [hunk 11]
```diff
@@ -441,6 +446,8 @@ dependencies {
     implementation 'com.vanniktech:emoji-material:0.7.0'
     implementation 'com.vanniktech:emoji-google:0.7.0'
 
+    implementation 'im.dlg:android-dialer:1.2.5'
```
Comment (bmarty) [inline vector/build.gradle:449]: Please add an entry in the file `open_source_licenses.html` for this lib

Comment (ganfra) [reply vector/build.gradle:449]: Weird I think I already did that 🤔

Comment (ganfra) [reply vector/build.gradle:449]: Yes it's already in

Comment (bmarty) [reply vector/build.gradle:449]: Ok, my bad, thanks

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/summary/RoomSummaryConstants.kt [hunk 12]
```diff
@@ -20,11 +20,15 @@ import org.matrix.android.sdk.api.session.events.model.EventType
 
 object RoomSummaryConstants {
 
+    /**
+     *
+     */
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/summary/RoomSummaryConstants.kt:25]: Thanks for this useful comment :p

**Review by bmarty — APPROVED** [review]
Let's merge that huge PR to develop!

**Comment by indefini** [comment]
it is already merged so not sure if my comment will be answered but, why version became a string here?
is it still possible to use version 0? I had something working in december and it is not working anymore.
The spec is not merged yet, but this is already implemented... shouldn't it be the other way around? (and then other clients can implement it too?)

**Comment by indefini** [comment]
> It is already merged so not sure if my comment will be answered but, why version became a string here?
ok I read the spec pull request there is discussion about it

**Pull request merged**